### DISCUSSION TITLE
MappedAsDataclass uses standard values ​​as explicit... why?

### DISCUSSION THREAD

**Discussion by nleononline** [sqlalchemy/sqlalchemy#12165] [Usage Questions]
Python: 3.12
SQLalchemy: 2.0.36

-----

A basic declarative class (without MappedAsDataclass) will ignore all values ​​that are not passed in explicitly. 

However, if we add MappedAsDataclass - this forces us to specify `mapped_column(default=None)` for the code to work at all. SQLAlchemy thinks that we passed the standard values ​​explicitly, which of course is not the case.

### Standart model:
```python
UserModel(Base)
    __tablename__ = user
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None]

user = UserModel(id=1, name="Jon", status="active")
session.merge(user)
session.commit()

user = UserModel(id=1, name="John")  << changed
session.merge(user)
session.commit()
```

**Database:** 

| id | name | status    |
| -- | ----- | -------- | 
|  1  | John | active   |

SQLAlchemy only updated the values ​​that were passed explicitly. In the given example field the **name**.

-------

### add MappedAsDataclass:
```python
UserMappedModel(MappedAsDataclass, Base)
    __tablename__ = user_mapped
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default=None)

user = UserMappedModel(id=1, name="Jon", status="active")
session.merge(user)
session.commit()

user = UserMappedModel(id=1, name="John")  << changed
session.merge(user)
session.commit()
```

**Database:** 

| id | name | status    |
| -- | ----- | -------- | 
|  1  | John | null        |

SQLAlchemy replaced the **name** field with **None**, although we did not specify this explicitly.

-----

A logic question arises: why does the library suddenly change behavior (once again) and how to deal with it? 

It just so happens that I cannot get the full object via session.get, because will be creating too much load. In this case session.merge() works perfectly for me, however with MappedAsDataclass I suddenly get different behavior.

**Comment by CaselIT** [comment]
Hi,

> A logic question arises: why does the library suddenly change behavior (once again) and how to deal with it?

From sqlalchemy point of view that are the same, meaning that status has the value set by the dataclass. That's the same as if you were passing `status=None` in the init. If you don't want that behaviour tell `MappedAsDataclass` to not generate the init

**Reply by zzzeek** [reply]
we have another discussion somewhere where we propose having some new value like "DONTSET" that can be set as a dataclass default.   still seems like we should move forward on that.  not sure if we made an issue out of it

**Reply by CaselIT** [reply]
IIRC that was the use case for dontset. It seems to be here: https://github.com/sqlalchemy/sqlalchemy/discussions/9383 that's linked here https://github.com/sqlalchemy/sqlalchemy/issues/9410, but I don't think there is an issue for it

NOTE that is should be supported also in the relationships

**Reply by zzzeek** [reply]
we really should do it because "notset" is certainly a valid state in a SQLAlchemy object that is distinct from None and it has its own meaning as we see here and elsewhere

**Reply by CaselIT** [reply]
it likely makes sense to target 2.1 so if we need to do some incompatible change it's ok to do

**Reply by zzzeek** [reply]
Im thinking it can be in 2.0 if we plaster it with "ALPHA - CAN CHANGE AT ANY TIME" warnings

**Reply by CaselIT** [reply]
as you prefer. I guess if it's a new behaviour that an user need to specifically opt into that's fine too.

**Reply by zzzeek** [reply]
i think one area that we have a choice to make is if you have an object with state and you say, `myobject.foo = NO_VALUE` - does that expire the attribute or just do nothing?  I vote for the latter

**Reply by CaselIT** [reply]
I guess the behaviour wanted is
```py
thing = session.get(Thing, 1)
assert thing.foo == 'bar'
thing.foo = NO_VALUE
assert thing.foo == 'bar'
```
and if `foo` is expired it stays expired, so
```py
thing = session.get(Thing, 1)
session.expire(thing, 'foo')
assert 'foo' in sa.inspect(x).expired_attributes
thing.foo = NO_VALUE
assert 'foo' in sa.inspect(x).expired_attributes
```
In the above example either behaviour is fine.
Not sure if before `assert thing.foo == 'bar'` in the first example 'foo' should be in the `expired_attributes` or not. I trust your insight here if you prefer for not.

**Reply by CaselIT** [reply]
I've created https://github.com/sqlalchemy/sqlalchemy/issues/12168 so let's continue there

**Reply by nleononline** [reply]
> Я создал [#12168,](https://github.com/sqlalchemy/sqlalchemy/issues/12168) так что давайте продолжим там.

I think adding a magic attribute by the user is not what is needed. As you can see in my example above - the basic models work correctly without the magic attributes I specified. 

There may be technical magic attributes, but the user should not specify them, because the behavior should be the same regardless of the model implementation.

I think there should be an exclude_unset mechanism, like in pydantic, but which works automatically, without having to call any method.

**Reply by zzzeek** [reply]
> There may be technical magic attributes, but the user should not specify them, because the behavior should be the same regardless of the model implementation.

the behavior cannot be the same for a dataclass without the advent of a DONTSET kind of value, since the contract of the dataclass is being honored here, the attribute is populated with None.  That None is then correctly transmitted in the merge() step.      With a DONTSET attribute, you would in fact see the value ``None`` if you tried to access the "not set" attribute on the object.

> I think there should be an exclude_unset mechanism, like in pydantic, but which works automatically, without having to call any method.

I dont use pydantic, but the DONTSET constant proposed is exactly such a mechanism.

**Reply by nleononline** [reply]
> > Могут существовать технические магические атрибуты, но пользователь не должен их указывать, поскольку поведение должно быть одинаковым независимо от реализации модели.
> 
> поведение не может быть таким же для dataclass без появления типа значения DONTSET, поскольку здесь соблюдается контракт dataclass, атрибут заполняется None. Затем этот None корректно передается на этапе merge(). С атрибутом DONTSET вы фактически увидите значение, `None`если попытаетесь получить доступ к атрибуту "not set" объекта.
> 
> > Я думаю, что должен быть механизм exclude_unset, как в pydantic, но который работает автоматически, без необходимости вызывать какой-либо метод.
> 
> Я не использую Pydantic, но предложенная константа DONTSET — это именно такой механизм.

I think that this should be added to mapped_column or default settings model, and not explicitly specified in the attribute, because this will pollute the code.

```python
UserMappedModel(MappedAsDataclass, Base)
    __tablename__ = user_mapped
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default=None, exclude_unset=True)  << NEW
```
or 

```python
UserMappedModel(MappedAsDataclass, Base)
    __tablename__ = user_mapped
    __table_args__ = {"exclude_unset": True}  << NEW
    
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default=None)
```

This approach minimizes the amount of code without having to assign a DONTSET value to each attribute each time.

it's bad example:
```python
user = UserMappedModel(id=1, name="John", status=NO_VALUE)
```

Ideally, this setting should be used **when executing session methods**. We do not violate the dataclass contract, but exclude unset fields before inserting them into the session.

Or you can add a method to the model as in pydantic and exclude implicit values ​​before adding the object to the session (but I like this option the least):
```python
session.merge(user.exclude(unset=True))
```

**Reply by zzzeek** [reply]
so this code IMO doesnt actually make any sense:

    mapped_column(default=None, exclude_unset=True)

What does "default" here indicate?   How is the above code different in any way from code that says this:

    mapped_column(default="This is a string that will never be used", exclude_unset=True)

Since exclude_unset means, "don't use the default".  This creates ambiguity and needless verbosity in the API

Additionally, the SQLAlchemy dataclass implementation **uses the plain Python dataclass constructor**, so there is also no implementation hook for this "exclude_unset" parameter, unless it behind the scenes changed the value of "default=None" to be...."default=UNSET" before setting up the constructor, where the proposal is to allow this directly.

> session.merge(user.exclude(unset=True))

as mentioned before, if you set `default=None` on a field of your dataclass, **that value is set**, so "unset" means nothing here

**Reply by CaselIT** [reply]
> There may be technical magic attributes, but the user should not specify them, because the behavior should be the same regardless of the model implementation.

That's not possible. The default constructor of a class is as follows:
```py
def __init__(self, **kw):
  for key in kw:
    setattr(self, key, kw[key]
```
If you don't specify status then nothing is set.

Then you add dataclass the init will be generated by the dataclass module, so that will be:
```py
def __init__(self, id, name, status=None):
    self.id = id
    self.name = name
    self.status = status
```

There is no unset attribute here. This is behavior of dataclasses, and we don't want to change it, since it's not appropriate for sqlalchemy to make decision about an builtin feature.

Currently your option is to avoid generating the init with dataclasses, so that the default constructor will be used.

What that issue proposes is a way of indicating that the default should be "do not set status" by doing
```py
class Model(MappedAsDataclass, Base)
    __tablename__ = user_mapped
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default=NO_VALUE)
```

**Reply by nleononline** [reply]
> так что этот код, по моему мнению, на самом деле не имеет никакого смысла:
> 
> ```
> mapped_column(default=None, exclude_unset=True)
> ```
> 
> Что здесь означает "default"? Чем вышеприведенный код отличается от кода, который говорит следующее:
> 
> ```
> mapped_column(default="This is a string that will never be used", exclude_unset=True)
> ```
> 
> Так как exclude_unset означает "не использовать значение по умолчанию". Это создает двусмысленность и ненужную многословность в API
> 
> Кроме того, реализация класса данных SQLAlchemy **использует простой конструктор класса данных Python** , поэтому для этого параметра «exclude_unset» также нет хука реализации, если только он не изменил значение «default=None» на... «default=UNSET» перед настройкой конструктора, в то время как предлагается разрешить это напрямую.
> 
> > сеанс.объединить(пользователь.исключить(снять=Истина))
> 
> как упоминалось ранее, если вы устанавливаете значение `default=None`для поля вашего класса данных, **это значение устанавливается** , поэтому «не задано» здесь ничего не означает

We don’t just have a dataclass object, it’s actually also a model of alchemy, right? 

in this case default=None is the dataclass value, and exclude_unset is a setting for the alchemy model.

We don’t have a pure dataclass, we have a mixture of alchemy (mapped_colum, realtionship).

-----

As for default=UNSET, there are a few oddities here. 

Firstly, we have a dataclass and can work with it in code, so we may need standard values ​​(otherwise why use dataclass?)

In this case, we cannot specify default=UNSET, because somewhere we need default=None, and somewhere default=False, or default=Status.ACTIVE. 

In this case, using default=UNSET does not really suit us.

**Reply by nleononline** [reply]
> > Могут существовать технические магические атрибуты, но пользователь не должен их указывать, поскольку поведение должно быть одинаковым независимо от реализации модели.
> 
> Это невозможно. Конструктор класса по умолчанию выглядит следующим образом:
> 
> ```python
> def __init__(self, **kw):
>   for key in kw:
>     setattr(self, key, kw[key]
> ```
> 
> Если статус не указан, то ничего не установлено.
> 
> Затем вы добавляете dataclass, init будет сгенерирован модулем dataclass, так что это будет:
> 
> ```python
> def __init__(self, id, name, status=None):
>     self.id = id
>     self.name = name
>     self.status = status
> ```
> 
> Здесь нет атрибута unset. Это поведение dataclasses, и мы не хотим его менять, поскольку sqlalchemy не подходит для принятия решения о встроенной функции.
> 
> В настоящее время вы можете избежать генерации init с помощью dataclasses, чтобы использовался конструктор по умолчанию.
> 
> В этой проблеме предлагается способ указать, что по умолчанию должно быть «не устанавливать статус», выполнив следующее:
> 
> ```python
> class Model(MappedAsDataclass, Base)
>     __tablename__ = user_mapped
>     id: Mapped[int] = mapped_column(primary_key=True)
>     name: Mapped[str]
>     status: Mapped[str | None] = mapped_column(default=NO_VALUE)
> ```

It's probably worth separating the behavior of dataclass from its use in session. 

I described this above - we need default values ​​for the dataclass object to work with it in code, but since... we did not set the values ​​explicitly - they should be ignored when working in session. 

This is the behavior you initially expect when working with alchemy (otherwise why would you need MappedAsDataclass?)

Maybe it's worth trying to make the TypeVar value for default - Default[...]? 

We leave default in the dataclass, and if the user explicitly specifies a value, it will no longer be a TypeVar type, and when working with a session, we can recognize which values ​​were explicitly specified and which were not.

```python
UserMappedModel(MappedAsDataclass, Base)
    __tablename__ = user_mapped
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default=None)
    
user = UserMappedModel(id=1, name="Name")
print(type(user.status))
> Default[None]
```

**Reply by CaselIT** [reply]
> I described this above - we need default values ​​for the dataclass object to work with it in code, but since... we did not set the values ​​explicitly - they should be ignored when working in session.

Then don't use the dataclass init.
If you use it there is no difference between using `Model(id=1, name='n', status=None)` and `Model(id=1, name='n')` once you have the instance.

Use instead
```py
class Model(MappedAsDataclass, Base, init=False)
    __tablename__ = user_mapped
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default=None)
```

**Reply by nleononline** [reply]
> > Я описал это выше — нам нужны значения по умолчанию для объекта dataclass, чтобы работать с ним в коде, но поскольку... мы не задали значения явно — их следует игнорировать при работе в сеансе.
> 
> Тогда не используйте dataclass init. Если вы используете его, то нет никакой разницы между использованием `Model(id=1, name='n', status=None)`и `Model(id=1, name='n')`как только у вас есть экземпляр.
> 
> Используйте вместо этого
> 
> ```python
> class Model(MappedAsDataclass, Base, init=False)
>     __tablename__ = user_mapped
>     id: Mapped[int] = mapped_column(primary_key=True)
>     name: Mapped[str]
>     status: Mapped[str | None] = mapped_column(default=NO_VALUE)
> ```

How do you get `default=Status.ACTIVE` in this example? This example forces you to abandon all default values ​​and specify them explicitly... which by the way does not work if `init=False` is specified and bonus warning "Unexpected argument" in IDE

**Reply by CaselIT** [reply]
sorry there was a typo in my last example here https://github.com/sqlalchemy/sqlalchemy/discussions/12165#discussioncomment-11512558 that I have corrected. The default for status should be `None` like in your opening post

**Reply by nleononline** [reply]
@CaselIT
It seems to me that the Default type https://github.com/sqlalchemy/sqlalchemy/discussions/12165#discussioncomment-11512506 can be a solution to this problem.

**Reply by nleononline** [reply]
> извините, в моем последнем примере здесь была опечатка [#12165 (ответ в теме),](https://github.com/sqlalchemy/sqlalchemy/discussions/12165#discussioncomment-11512558) я исправил. Значение по умолчанию для последствий должно быть `None`таким, как в вашем начальном посте.

not work + warning IDE "Unexpected argument" :)

**Reply by CaselIT** [reply]
> @CaselIT It seems to me that the Default type [#12165 (reply in thread)](https://github.com/sqlalchemy/sqlalchemy/discussions/12165#discussioncomment-11512506) can be a solution to this problem.

that would be something else though, not a dataclass. The objective of sqlalchemy dataclass support is to not change the behaviour of dataclasses.

> not work + warning IDE "Unexpected argument" :)

seems that you need to invert Base and MappedAsDataclass. That seems a bug of the type checkers.

The following works and types check. The model init is understood by the my ide as `class Model(**kw: Any)`
```py
import sqlalchemy as sa
from sqlalchemy.orm import (
    MappedAsDataclass,
    DeclarativeBase,
    mapped_column,
    Mapped,
)

class Base(DeclarativeBase):
    pass

class Model(Base, MappedAsDataclass, init=False):
    __tablename__ = "name"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default=None)

m1 = Model(id=1, name="foo")
m2 = Model(id=1, name="bar", status=None)
print("m1", m1.__dict__, m1.status)
print("m2", m2.__dict__, m2.status)
```

If you want you can even trick the type checker doing the following.
```py
import dataclasses
import sqlalchemy as sa
from sqlalchemy.orm import (
    MappedAsDataclass,
    DeclarativeBase,
    mapped_column,
    Mapped,
)

class Base(DeclarativeBase):
    pass

def my_dataclass(cls, **kwargs):
    kwargs.setdefault("init", False)
    return dataclasses.dataclass(cls, **kwargs)

class Model(Base, MappedAsDataclass, dataclass_callable=my_dataclass):
    __tablename__ = "name"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default=None)

m1 = Model(id=1, name="foo")
m2 = Model(id=1, name="bar", status=None)
print("m1", m1.__dict__, m1.status)
print("m2", m2.__dict__, m2.status)
```

in the above example my editor is giving the fully typed init:
```py
class Model(
    id: SQLCoreOperations[int] | int,
    name: SQLCoreOperations[str] | str,
    status: SQLCoreOperations[str | None] | str | None = None
)
```

**Reply by nleononline** [reply]
> > @CaselITМне кажется, что тип по умолчанию [#12165 (ответ в теме)](https://github.com/sqlalchemy/sqlalchemy/discussions/12165#discussioncomment-11512506) может стать решением этой проблемы.
> 
> Это было бы что-то другое, а не dataclass. Цель поддержки dataclass в sqlalchemy — не менять поведение dataclass.
> 
> > не работает + предупреждение IDE "Неожиданный аргумент" :)
> 
> кажется, вам нужно инвертировать Base и MappedAsDataclass. Это похоже на ошибку проверки типов.
> 
> Следующие работы и типы проверяются. Модель init понимается моей ide как`class Model(**kw: Any)`
> 
> ```python
> import sqlalchemy as sa
> from sqlalchemy.orm import (
>     MappedAsDataclass,
>     DeclarativeBase,
>     mapped_column,
>     Mapped,
> )
> 
> class Base(DeclarativeBase):
>     pass
> 
> class Model(Base, MappedAsDataclass, init=False):
>     __tablename__ = "name"
>     id: Mapped[int] = mapped_column(primary_key=True)
>     name: Mapped[str]
>     status: Mapped[str | None] = mapped_column(default=None)
> 
> m1 = Model(id=1, name="foo")
> m2 = Model(id=1, name="bar", status=None)
> print("m1", m1.__dict__, m1.status)
> print("m2", m2.__dict__, m2.status)
> ```
> 
> Если хотите, вы можете даже обмануть средство проверки типов, выполнив следующее.
> 
> ```python
> import dataclasses
> import sqlalchemy as sa
> from sqlalchemy.orm import (
>     MappedAsDataclass,
>     DeclarativeBase,
>     mapped_column,
>     Mapped,
> )
> 
> class Base(DeclarativeBase):
>     pass
> 
> def my_dataclass(cls, **kwargs):
>     kwargs.setdefault("init", False)
>     return dataclasses.dataclass(cls, **kwargs)
> 
> class Model(Base, MappedAsDataclass, dataclass_callable=my_dataclass):
>     __tablename__ = "name"
>     id: Mapped[int] = mapped_column(primary_key=True)
>     name: Mapped[str]
>     status: Mapped[str | None] = mapped_column(default=None)
> 
> m1 = Model(id=1, name="foo")
> m2 = Model(id=1, name="bar", status=None)
> print("m1", m1.__dict__, m1.status)
> print("m2", m2.__dict__, m2.status)
> ```
> 
> В приведенном выше примере мой редактор выдает полностью типизированный init:
> 
> ```python
> class Model(
>     id: SQLCoreOperations[int] | int,
>     name: SQLCoreOperations[str] | str,
>     status: SQLCoreOperations[str | None] | str | None = None
> )
> ```

It doesn't work the way you expect.

Once again to the problem:
1. Dataclass MUST have default values ​​to work in code.
2. Default values ​​MUST NOT go into session.merge.

init=False excludes the default value, which means you cannot rely on this dataclass in your code and there is no point in using it...

```python
class Base(DeclarativeBase):
    pass
    
def my_dataclass(cls, **kwargs):
    kwargs.setdefault("init", False)
    return dataclasses.dataclass(cls, **kwargs)  # noqa

class CreatedUpdatedMixin(MappedAsDataclass, kw_only=True):
    created_at: Mapped[datetime.datetime] = mapped_column(
        init=False,
        server_default=func.current_timestamp(),
        insert_default=func.now()
    )
    updated_at: Mapped[datetime.datetime] = mapped_column(
        default_factory=datetime.datetime.now,
        server_default=func.current_timestamp(),
        insert_default=func.now(),
        onupdate=func.now(),
        server_onupdate=func.current_timestamp()
    )

class User(Base, CreatedUpdatedMixin, kw_only=True, dataclass_callable=my_dataclass):
    __tablename__ = 'user'

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default="ACTIVE")
    is_active: Mapped[bool] = mapped_column(default=False)


user = User(id=1, name="John")
print(user)
> User(
    created_at=None, 
    datetime.datetime(2024, 12, 10, 0, 52, 47, 643649),
    id=1, 
    name='John', 
    status=None,  << NOT DEFAULT VALUE
    is_active=None   << NOT DEFAULT VALUE
)

print(user.__dict__)
> {
    '_sa_instance_state': <sqlalchemy.orm.state.InstanceState object at 0x106ab6ab0>, 
    'updated_at': datetime.datetime(2024, 12, 10, 0, 52, 47, 643649), 
    'id': 1, 
    'name': 'John'
} 
```

Now there is no right solution out of the alchemy “box”. Either default values ​​go into the database, or default values ​​are not in the dataclass. 

Correctly, there should be a separation between the object you work with in the code and the data that is passed to the session

**Reply by zzzeek** [reply]
> in this case default=None is the dataclass value, and exclude_unset is a setting for the alchemy model.

If we *do* use the default then I dont see what "exclude_unset" does unless you are proposing this is an obscure keyword that does a very specific thing when `session.merge()` is used which would be a poor API pattern to introduce, we dont add little bells and switches here and there to suit very specific quirks, we try to create APIs that make sense in broad strokes and allow quirks to be possible via generalized means

**Reply by nleononline** [reply]
> > в этом случае default=None — это значение класса данных, а exclude_unset — это настройка для модели алхимии.
> 
> Если мы _используем_ значение по умолчанию, то я не понимаю, что делает «exclude_unset», если только вы не предлагаете, что это непонятное ключевое слово, которое при `session.merge()`использовании делает очень специфическую вещь, что было бы плохим шаблоном API для внедрения. Мы не добавляем маленькие колокольчики и переключатели здесь и там, чтобы соответствовать очень специфическим особенностям. Мы пытаемся создавать API, которые имеют смысл в общих чертах и ​​позволяют реализовать особенности с помощью обобщенных средств.

excludes default values ​​before executing session.add and session.merge()

A very specific feature is that the models behave differently, although MappedAsDataclass is an out-of-the-box solution to alchemy and should definitely take into account all the basic and obvious scenarios for using objects with alchemy session methods.

**Reply by zzzeek** [reply]
> excludes default values ​​before executing session.add and session.merge()

this is really not how SQLAlchemy does things.    There is no way to distinguish between a `None` in the dictionary that came from the dataclass default, and a `None` that you set there.    Especially for `add()`, it's not possible to "ignore" a value when you `add()`, the object is not mutated.

This is all a very strange line of feature request that nobody has ever asked for before and it implies some entirely different way of operating which is not something SQLAlchemy can just whip up as a whole new way of working for whatever specific issues you might be having

**Reply by nleononline** [reply]
> > исключает значения по умолчанию перед выполнением session.add и session.merge()
> 
> SQLAlchemy делает это совсем не так. Нет способа отличить a `None`в словаре, который пришел из dataclass default, и a `None`, который вы там установили. Особенно для `add()`, невозможно «игнорировать» значение, когда вы `add()`, объект не мутировал.
> 
> Это очень странная строка запроса на функцию, о которой никто никогда раньше не просил, и она подразумевает совершенно иной способ работы, который SQLAlchemy не может просто так придумать как совершенно новый способ работы для решения конкретных проблем, с которыми вы можете столкнуться.

https://github.com/sqlalchemy/sqlalchemy/discussions/12165#discussioncomment-11513942 - simple solution to the problem

**Comment by zzzeek** [comment]
truth be told the bigger problem here is just the use of `merge()` in the first place.    `merge()` is not really that useful of a method since it has to make a lot of big assumptions as we see here, I've never found myself actually using it outside of the caching use case.   If I wrote SQLAlchemy today I'd probably not have implemented this method, and instead would have made something more generalized like an object-walking API or something like that people could use to write their own copy routines.

maybe just try to set the persisted state you are looking for on objects directly and/or implement your own copy routines

**Reply by nleononline** [reply]
> По правде говоря, большая проблема здесь заключается в использовании `merge()`в первую очередь. `merge()`на самом деле не такой уж полезный метод, поскольку он должен делать много больших предположений, как мы видим здесь, я никогда не обнаруживал, что использую его вне случая использования кэширования. Если бы я писал SQLAlchemy сегодня, я бы, вероятно, не реализовал этот метод, а вместо этого сделал бы что-то более обобщенное, например API обхода объектов или что-то в этом роде, что люди могли бы использовать для написания своих собственных процедур копирования.
> 
> может быть, просто попробуйте установить постоянное состояние, которое вы ищете, непосредственно для объектов и/или реализовать собственные процедуры копирования

session.merge() is great for merging objects when updating if you don't want to fetch a lot of data from the database. 

This works well for high-load applications.

**Reply by zzzeek** [reply]
merge is just a way to copy data from one object to another which you can do without merge() as well

what i am saying is it's not worth it to start building esoteric rules and switches into merge(), people will want one field to copy no matter what, another field to copy but only if it's not one of these three values, in your case you're asking that we not only have to compare an observed value to the "default" field but we also have to somehow know that this value *came from the default and was not passed explicitly*, which is completely not possible without some significant major complexity added (this is if we dont use my simple DONTSET constant which I still dont see what the problem is), this is all out of scope for where merge() came from

**Reply by nleononline** [reply]
> merge — это просто способ скопировать данные из одного объекта в другой, что можно сделать и без merge()
> 
> Я говорю о том, что не стоит начинать встраивать эзотерические правила и переключатели в merge(), люди захотят скопировать одно поле, несмотря ни на что, и другое поле, но только если оно не является одним из этих трех значений. В вашем случае вы просите, чтобы мы не только сравнивали наблюдаемое значение с полем «по умолчанию», но и каким-то образом узнавали, что это значение _взято из поля по умолчанию и не было передано явно_ , что совершенно невозможно без существенного усложнения (это если мы не используем мою простую константу DONTSET, в чем проблема, я все еще не понимаю). Все это выходит за рамки того, откуда взялась merge().

You could use something like this:
```python
DefaultArg = TypeVar("DefaultArg")


class Default(Generic[DefaultArg]):
    def __init__(self, value: DefaultArg):
        self.value = value

    def __repr__(self) -> DefaultArg:
        return f"{self.value}"


def is_default(value: Any) -> bool:
    return isinstance(value, Default)

    
class User(Base, CreatedUpdatedMixin, kw_only=True):
    __tablename__ = 'user'

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[str | None] = mapped_column(default=Default("ACTIVE"))  << Default TypeVar
    is_active: Mapped[bool] = mapped_column(default=Default(False))  << Default TypeVar
    

user = User(id=1, name="John")
print(user.status)
> "ACTIVE"
print(is_default(user.status))
> True
```

This will allow you not to interfere with the dataclass, but will make it clear to sqlalchemy that this is a default attribute and should be excluded from session methods and implement the same behavior of the model with implicit values when using MappedAsDataclass. 

```python
user2 = User(id=1, name="John", status="STOP")
print(user2.status)
> "STOP"
print(is_default(user2.status))
> False

```

If the user specifies the attribute manually, it will not be specified via Default and we will pass it to the session methods.

The only thing I haven’t completed here is the correct accounting of types, so that when calling `print(user.status is str)` it returns `True`, just as for `print(instatnse(user.status, str))` it returns `True`

**Reply by CaselIT** [reply]
> This will allow you not to interfere with the dataclass,

It doesn't type check since `Default(False)` is not a bool. Also you can't use that as a bool or whatever type, so it's not a solution.

Personally for this problem you are probably better served by two classes, one that has all your logic in it, then you have a metod to go from your class to the sqlalchemy model instance, doing whatever you like with the values there

**Reply by nleononline** [reply]
> > Это позволит вам не вмешиваться в класс данных,
> 
> Он не проверяет тип, так как `Default(False)`не является bool. Также вы не можете использовать его как bool или какой-либо другой тип, так что это не решение.
> 
> Лично для решения этой проблемы вам, вероятно, лучше подойдут два класса: один, содержащий всю вашу логику, а затем у вас есть метод перехода от вашего класса к экземпляру модели sqlalchemy, делая с имеющимися там значениями все, что вам угодно.

I just suggested one of the design options to solve the problem. You can refine this solution, or you can think about something else. I'm just offering options; it's obvious that they need to be improved. It seems like we found out early on that there was a problem and it was worth introducing some changes to alchemy so that default values ​​could be excluded. The question remains in the implementation, so why do you write that I should abandon the alchemy model in favor of “another class”? 

What's the point of MappedAsDataclass if it's not fully integrated into alchemy?

**Reply by CaselIT** [reply]
> I just suggested one of the design options to solve the problem. You can refine this solution, or you can think about something else. I'm just offering options; it's obvious that they need to be improved

Sorry but I disagree, you keep pushing your container solution dismissing anything else even after pointing out that it's incompatible with how dataclasses are typed

> The question remains in the implementation, so why do you write that I should abandon the alchemy model in favor of “another class”?

Because you want a very strange behaviour yo happen. You want some fields to be populated but you don't want them to be seen by sqlalchemy. That's not how sqlalchemy models or dataclasses works, so it's very likely that any change implemented will not satisfy your very peculiar use case.

> What's the point of MappedAsDataclass if it's not fully integrated into alchemy?

It is fully integrated. What you are looking for is not how dataclasses work.

**Reply by nleononline** [reply]
> > Я просто предложил один из вариантов дизайна для решения проблемы. Вы можете доработать это решение, а можете подумать о чем-то другом. Я просто предлагаю варианты, очевидно, что их нужно улучшать
> 
> Извините, но я не согласен. Вы продолжаете продвигать свое решение с контейнером, отвергая все остальное, даже указав, что оно несовместимо с типом классов данных.
> 
> > Вопрос остается в реализации, так почему же вы пишете, что я должен отказаться от модели алхимии в пользу «другого класса»?
> 
> Потому что вы хотите очень странного поведения, вы случаетесь. Вы хотите, чтобы некоторые поля были заполнены, но вы не хотите, чтобы они были видны sqlalchemy. Это не то, как работают модели или классы данных sqlalchemy, поэтому весьма вероятно, что любое реализованное изменение не удовлетворит ваш очень странный вариант использования.
> 
> > В чем смысл MappedAsDataclass, если он не полностью интегрирован в алхимию?
> 
> Он полностью интегрирован. То, что вы ищете, — это не то, как работают классы данных.

https://github.com/sqlalchemy/sqlalchemy/discussions/12165#discussion-7641197 read the very first message again and you will find different behavior depending on whether MappedAsDataclass is used or not. 

Different behavior is not correct. We decided yesterday that this was a problem. 

I have nothing more to add.

**Reply by CaselIT** [reply]
> Different behavior is not correct. We decided yesterday that this was a problem.

The different behaviour is how a dataclass behave. I mentioned it multiple times. My initial comment suggests not using them, so don't understand why you keep mentioning it

**Comment by zzzeek** [comment]
Here's a generalized recipe for ignoring m2o relationship as None for the 2.0 series until #12168 is available: https://github.com/sqlalchemy/sqlalchemy/issues/12168#issuecomment-2892092715
